include/tvm/arith/linear_system.h

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file tvm/arith/linear_system.h
+ * \brief Linear system data structures and solvers
+ */
+#ifndef TVM_ARITH_LINEAR_SYSTEM_H_
+#define TVM_ARITH_LINEAR_SYSTEM_H_
+
+#include <tvm/ir/expr.h>
+#include <tvm/tir/expr.h>
+#include <unordered_map>
+#include <vector>
+
+namespace tvm {
+namespace arith {
+
+using tir::Var;
+using tir::VarNode;
+using tir::IterVar;
+
+/*!
+ * \brief Represent a linear system including variables, their ranges and
+ *        the linear relations between them (either equations or inequalities)
```

Comment:
I think it's worth mentioning that the system is integer.

Also I'm not sure if LinearSystem is a good name, because it may be useful to have non-linear (in)equalities in `relations` (which doesn't break the algorithms that work on the linear part).

Reply:
good suggestion. I have several proposals, how about Constraints, RelationSet, IntRelationSet, or InequalitySet (if we make term "inequality" to also include equations)?

Reply:
IntegerSet or IntegerSystem maybe?

Reply:
sounds good. There's IntSet in tvm, so I'll do IntegerSystem.

Reply:
think it a bit more, should we use a more general name? as in solver, we will check the relation and ignore the non-integer equations, which implies it is valid to contain non-integer relations inside. @sergei-grechanik

Reply:
If variables are integer, I think it's still ok to call it IntegerSystem. But if we allow non-integer variables (I doubt it is a good idea), then I don't know, Set and System sound too unspecific, Constraints may be ok.